Monehin/pulse-backend
# Register user with invite cohort-program

## What?
Add Cohort-Program assignment in Invite
## Why?
So on registration, the user is automatically assigned to the applicable program schedule
## How?
- Add a relationship between Invites and Cohort-Program schedules
- Create a dynamic default Cohort-Program for the period of invitation in the Invite
- Check invite for program schedules on registration
     - If provided, assign registration target to specified schedule
     - If not express, compute the current Bootcamp cycle as default and assign
## Testing?
- Checkout to the branch _ft-add-cohort-program-in-invite_
- Run your development server
- Setup the requisite test data in your local database
  - Ensure all Cohort-Programs schedules have at least a start date, especially if you have created cohorts or programs on this branch
- In your Strapi admin panel, create two invites, one with a Cohort-Program selected, another without
- Either register each of the two users on postman on the route _POST: localhost:1337/auth/local/register_ .
  - The request body should be a JSON object with a minimum of username, email, and password properties and their values.
  - The email used must have an existing invite
- Or start your local front end and enter localhost:<your port>/register in your address bar
  - Complete the registration process 
- In turns, open each user on your Strapi UI or postman and see that:
     - The user invited without a Cohort-Program specified is assigned a default schedule
     - The user invited with a Cohort-Program specified is assigned to the schedule assigned in the invite
     - Try changing the scheduled start date of Cohort-Programs and see that the default Cohort-Program assigned is always the current one - around the time of invitation
## Screenshots (optional)
![default schedule](https://slack-files.com/T02R3LKBA-F020U2LH8R1-0622da530e "Screenshot of registration with defaults")
## Anything Else?
- Creating a Cohort automatically creates and populates all applicable Cohort-Program combinations for the new Cohort and vice-versa.
  - For each automatically populated Cohort-Program, please edit and provide a start date, especially the Cohort-Bootcamp combinations.
  - This will aid in getting a default Cohort-Program if not explicitly specified in the invite
- Name fields have been added to Cohort-program-schedule and User-cohort-programs models so entries can be displayed by their names rather than indecipherable Ids
- The admin panel accordingly displays these relationship fields by their names instead of Id
- Names are automatically generated for named relationships like Cohort-Programs and User-Cohort-Programs following that order. This functionality has been made generic and can be extended to practically any relationship
- The Super Admin can override the automatic naming in favor of custom naming
- Additionally, defaults added for invitee role if not provided in invite and Invitee-cohort-program manager if not provided. The invitee-cohort-program manager defaults to the inviter, if of role Manager
